Partial artifact matching based test suite (#210)
hylo-lang/hylo-new: Tests/CompilerTests/README.md

A single-file test is compiled to a binary executable, just as if it was passed as an argument to `hc`.
A package test is built according to the configuration specified by its manifest.

## Testing without the standard library

You can add `//! no-std` on the first line of a single-file test to disable the loading of the standard library. We encourage using `no-std` when possible, as it makes the test significantly faster. 

## Stopping after a compilation stage

> Add the `//! stage:stage_name` manifest attribute to stop after a specific compilation stage. Valid stages are:

Reviewer: ```suggestion

Add `//! stage:stage_name` to stop the driver after a specific compilation stage. Valid stages are:
```

- `stage:parsing`
- `stage:typing`
- `stage:lowering`
- `stage:llvm`
- `stage:execution` (default); applicable only in positive tests.

## Expectation types

### Exit status

> Specify an expected exit code by `exit-status:42` manifest attribute. `0` is expected by default.

Reviewer: ```suggestion

Specify an expected exit code by `exit-status:42` manifest attribute. `0` is expected by default.
```

Asserted when `stage` is set to `execution` (default). 

### Textual Artifact Expectations

> You can add a `test-case.<artifact-tag>.expected` file besides your `test-case.hylo` to assert the contents of a compilation artifact. Valid artifacts are:

Reviewer: ```suggestion

Add a `test-case.<artifact-tag>.expected` file besides your `test-case.hylo` to assert the contents of a compilation artifact. Valid artifacts are:
```

- `*.raw-ir.expected`
- `*.transformed-ir.expected`
- `*.llvm-ir.expected`

An artifact is composed of a set of contiguous sections, delimited by empty lines. E.g. each function in Hylo IR and LLVM IR are their own sections.

It is sufficient to specify a subset of sections of the actual artifact. The expected section is matched with the observed section having the closest first line. Then they are compared for equality.

## Inspecting Intermediate Artifacts

> The intermediate compilation artifacts are saved on test failure as `*.observed` files besides your test case. To save these even when a test passes, you can temporarily set `artifactsAreSavedOnSuccess` to `true` in `CompilerTests.swift`.

Reviewer: ```suggestion

Intermediate compilation artifacts are saved on test failure as `*.observed` files besides the corresponding test case. Aou can temporarily set `artifactsAreSavedOnSuccess` to `true` in `CompilerTests.swift` to save these even when a test passes.
```

I think this option should be controlled by an environment variable.

Author: I considered that, but unfortunately environment variables can't be changed easily in IDEs without restarting them. I change this variable whenever I'm debugging something, so there needs to be a way to flip this flag easily. I don't think anyone would want to set this flag to true forever in their environment, as that can significantly slow down usual testing times due to saving files.
